### PR TITLE
Add Bias to Tensor

### PR DESCRIPTION
**Pull request by Semyon1104**

### CODE REVIEW

**Comment by Semyon1104** [comment]
Please look at the changes in the tensor, so that I understand whether it is worth changing the rest of the layers based on it or not?

File: include/layers/ConvLayer.hpp [hunk 1]
```diff
@@ -13,16 +13,19 @@ class ConvolutionalLayer : public Layer {
...
   ConvolutionalLayer() = default;
   ConvolutionalLayer(size_t step, size_t pads, size_t dilations,
-                     const Tensor& kernel) {
-    stride_ = step;
-    pads_ = pads;
-    dilations_ = dilations;
-    kernel_ = kernel;
-  }
+                     const Tensor& kernel,
+                     const Tensor& bias = Tensor())  // bias �� ���������
```
Comment (aobolensk) [inline include/layers/ConvLayer.hpp:22]: Please fix the comments

File: include/layers/Tensor.hpp [hunk 2]
```diff
@@ -1,5 +1,3 @@
-#pragma once
-
```
Comment (aobolensk) [inline include/layers/Tensor.hpp:2]: Why it was removed?

File: src/layers/ConvLayer.cpp [hunk 3]
```diff
@@ -94,4 +105,5 @@ void ConvolutionalLayer::run(const Tensor& input, Tensor& output) {
     }
   }
 }
-}  // namespace itlab_2023
+
+}  // namespace itlab_2023
```
Comment (aobolensk) [inline src/layers/ConvLayer.cpp:109]:
```suggestion
}  // namespace itlab_2023

```

File: src/layers/Tensor.cpp [hunk 4]
```diff
@@ -14,6 +14,16 @@ std::ostream& operator<<(std::ostream& out, const Tensor& t) {
...
   }
+
+  if (!t.get_bias().empty()) {
+    out << "Bias: ";
+    for (const auto& b : t.get_bias()) {
+      out << b << " ";
+    }
+    out << std::endl;
+  }
+
```
Comment (aobolensk) [inline src/layers/Tensor.cpp:25]: Remove

**Review by aobolensk — CHANGES_REQUESTED** [review]

File: include/layers/Tensor.hpp [hunk 5]
```diff
@@ -67,16 +60,62 @@ class Tensor {
...
   }
 
-  Tensor(const Tensor& t) = default;
+  Tensor(const std::vector<uint8_t>& a, const Shape& sh,
+         const std::vector<float>& bias)
+      : shape_(sh), values_(a), bias_(bias), type_(Type::kFloat) {
+    if (a.size() != SetRightTypeValues().size()) {
+      throw std::invalid_argument("Incorrect vector size given to Tensor");
+    }
+    if (bias.size() != shape_[1]) {
```
Comment (aobolensk) [inline include/layers/Tensor.hpp:69]: I guess, we need to check if shape_[1] exists before (check for shape.size())

File: include/layers/Tensor.hpp [hunk 6]
```diff
@@ -67,16 +60,62 @@ class Tensor {
...
+  auto begin() const { return values_.begin(); }
+
+  auto end() const { return values_.end(); }
+
+  size_t size() const {
+    size_t total_size = 1;
+    for (size_t i = 0; i < shape_.dims(); ++i) {
+      total_size *= shape_[i];
+    }
+    return total_size;
```
Comment (aobolensk) [inline include/layers/Tensor.hpp:116]: Consider using std::accumulate

**Review by aobolensk — APPROVED** [review]

**Pull request merged**